addons/account/models/account_bank_statement.py

```diff
@@ -1070,32 +1071,51 @@ def _prepare_reconciliation(self, lines_vals_list):
                 # Existing account.move.line.
                 to_browse_ids.append(vals.pop('id'))
                 to_process_vals.append(vals)
+                if any(x in vals for x in ('balance', 'amount_residual', 'amount_residual_currency')):
+                    special_case = False
             else:
                 # Newly created account.move.line from scratch.
                 line_vals = self._prepare_counterpart_move_line_vals(vals)
                 total_balance += line_vals['debit'] - line_vals['credit']
                 total_amount_currency += line_vals['amount_currency']
-
-                reconciliation_overview.append({
-                    'line_vals': line_vals,
-                })
+                reconciliation_overview.append({'line_vals': line_vals})
+                special_case = False
 
         # Step 2: Browse counterpart lines all in one and process them.
 
         existing_lines = self.env['account.move.line'].browse(to_browse_ids)
+
+        i = 0
         for line, counterpart_vals in zip(existing_lines, to_process_vals):
             line_vals = self._prepare_counterpart_move_line_vals(counterpart_vals, move_line=line)
             balance = line_vals['debit'] - line_vals['credit']
             amount_currency = line_vals['amount_currency']
+            i += 1
+
+            if i == len(existing_lines):
```

Comment:
I don't get this condition; feels weird. Also, i is probably not a very good name if your not incrementing the counter at the end of the loop. Maybe enumerate() would make this clearer ?